examples/Encrypted Inference-LinearRegression.ipynb

```diff
@@ -0,0 +1,1657 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #11.      mpc_model = model.share(session)
</pre><p>Related with the above comments.</p><p>In this function, seems that the <code>model</code>  and the <code>test_data</code>  belong to client (the one is running the notebook), but if I have the model and the and data... why would I want to encrypt both? (maybe I am not undersating the function).</p><p>Also, is it possible a more realistic situation where the model is already hosted and encrypted somewhere and I encrypt my data, run this <code>encrypted inference</code>  and get the result in plaintext?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/SyMPC/pull/276/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

Reply:
Right so here I would want to demonstrate only using the Virtual Workers feature of Syft. The more realistic scenario demonstrated with the duet will be for a future notebook.  But, for now just showing SyMPC's capabilities to perform encrypted inference with virtual workers.

Reply:
<p>You can send the data to the client and the use <code>MPCTensor</code></p>

Reply:
Right, seems good.

Reply:
@gmuraru Have to make this change. Don't merge yet.

Reply:
I tried to change it accordingly by:

`def inference(n_clients,protocol=None):
    
  # Get VM clients 
  parties=get_clients(n_clients)

  # Setup the session for the computation
  if(protocol):
     session = Session(parties = parties,protocol = protocol)
  else:
     session = Session(parties = parties)
        
  SessionManager.setup_mpc(session)

  #Send data to clients
  
  pointers = split_send(test_x,parties)

  #Encrypt model 
  mpc_model = model.share(session)

  #Encrypt test data
  #test_data=MPCTensor(secret=test_x, session = session)

  #Perform inference and measure time taken
  start_time = time.time()

  for ptr in pointers:
     print(ptr)
     enc_results = mpc_model(ptr)
  end_time = time.time()

  print(f"Time for inference: {end_time-start_time}s")

  #Get plaintext predictions
  predictions = enc_results.reconstruct()
  
  #Calculate Loss
  print("MSE Loss: ",criterion(predictions.reshape([-1]),test_y).item())
    
  return predictions`

Spit and send does this

`def split_send(data,parties):
    
    data_pointers = []
    
    split_size = int(len(data)/len(parties))
    for index in range(0,len(parties)):
        ptr=data[index*split_size:index*split_size+split_size].send(parties[index])
        data_pointers.append(ptr)
        
    return data_pointers`

Reply:
It ended up with this error 

```
<syft.proxy.torch.TensorPointer object at 0x7fecf971d400>
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-26-a5bbb142213b> in <module>
----> 1 predictions=inference(3,Falcon("semi-honest"))

<ipython-input-25-bbb75a2421aa> in inference(n_clients, protocol)
     27   for ptr in pointers:
     28      print(ptr)
---> 29      enc_results = mpc_model(ptr)
     30   end_time = time.time()
     31 

~/opt/anaconda3/envs/sympc/lib/python3.9/site-packages/syft/lib/torch/module.py in __call__(self, *args, **kwargs)
    171         self, *args: Union[List[Any], Tuple[Any, ...]], **kwargs: Dict[Any, Any]
    172     ) -> Any:
--> 173         return self.forward(*args, **kwargs)
    174 
    175     @property

<ipython-input-17-f64f170e7ce9> in forward(self, x)
      6 
      7     def forward(self, x):
----> 8         x = self.fc1(x)
      9         return x

~/Hrishikesh/Projects/SyMPC/src/sympc/module/nn/linear.py in forward(self, x)
     54             An MPCTensor that results by applying the layer specific operation on the input
     55         """
---> 56         res = x @ self.weight.t()
     57 
     58         if self.bias is not None:

~/opt/anaconda3/envs/sympc/lib/python3.9/site-packages/syft/ast/klass.py in run_class_method(__self, *args, **kwargs)
    138 
    139             # then we convert anything which isnt a pointer into a pointer
--> 140             pointer_args, pointer_kwargs = pointerize_args_and_kwargs(
    141                 args=downcast_args, kwargs=downcast_kwargs, client=__self.client
    142             )

~/opt/anaconda3/envs/sympc/lib/python3.9/site-packages/syft/ast/klass.py in pointerize_args_and_kwargs(args, kwargs, client)
    839         # check if its already a pointer
    840         if not isinstance(arg, Pointer):
--> 841             arg_ptr = arg.send(client, pointable=False)
    842             pointer_args.append(arg_ptr)
    843         else:

~/Hrishikesh/Projects/SyMPC/src/sympc/tensor/mpc_tensor.py in __getattribute__(self, attr_name)
    900             return functools.partial(approx_func, self)
    901 
--> 902         return object.__getattribute__(self, attr_name)
    903 
    904     def backward(self, gradient: Optional["MPCTensor"] = None) -> None:

AttributeError: 'MPCTensor' object has no attribute 'send'
```

Reply:
Any idea what could be the problem?